File: internal/handler/ocr.go
package handler

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"flowconvert/internal/config"
	"flowconvert/internal/service"
)

// OcrH handles OCR endpoints.
type OcrH struct {
	Cfg   *config.Config
	Store *FileStore
	Jobs  *OcrJobStore
}

var ocrExts = []string{"jpg", "jpeg", "png", "bmp", "tiff", "tif", "webp", "gif", "pdf"}

var ocrDownloadNames = map[string]string{
	"txt":    "识别结果.txt",
	"docx":   "识别结果.docx",
	"xlsx":   "识别结果.xlsx",
	"json":   "识别结果.json",
	"redbox": "红框标记图.png",
}

var ocrExportOrder = []string{"txt", "docx", "xlsx", "json", "redbox"}

func (h *OcrH) writeJSON(w http.ResponseWriter, status int, v interface{}) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(v)
}

func (h *OcrH) writeErr(w http.ResponseWriter, status int, msg string) {
	h.writeJSON(w, status, map[string]interface{}{"success": false, "error": msg})
}

func (h *OcrH) safeErr(w http.ResponseWriter, err error) {
	fmt.Fprintf(os.Stderr, "[OCR] error: %v\n", err)
	h.writeJSON(w, http.StatusUnprocessableEntity, map[string]interface{}{"success": false, "error": "识别失败，请稍后重试"})
}

func (h *OcrH) newTmp() (string, error) {
	dir := filepath.Join(h.Cfg.TmpDir, "ocr_"+service.NewID(8))
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return "", err
	}
	return dir, nil
}

func (h *OcrH) cleanupTmp(dir string) {
	_ = os.RemoveAll(dir)
}

// isMultipartForm reports whether the request body is multipart form data.
func isMultipartForm(r *http.Request) bool {
	ct := r.Header.Get("Content-Type")
	if i := strings.Index(ct, ";"); i >= 0 {
		ct = ct[:i]
	}
	return strings.HasPrefix(strings.ToLower(strings.TrimSpace(ct)), "multipart/")
}

// parseOcrBool accepts the truthy spellings browsers and curl send for a checkbox.
func parseOcrBool(v string) bool {
	switch strings.ToLower(strings.TrimSpace(v)) {
	case "1", "true", "yes", "on":
		return true
	}
	return false
}

// parseOcrOptions reads the option form fields; empty values keep their defaults
// so Normalize can validate them.
func parseOcrOptions(r *http.Request) service.OcrOptions {
	// ParseForm alone does not read multipart bodies, so the right parser has
	// to be picked based on the content type.
	if r.Form == nil {
		if isMultipartForm(r) {
			_ = r.ParseMultipartForm(32 << 20)
		} else {
			_ = r.ParseForm()
		}
	}
	form := r.Form
	formats := []string{"txt"}
	if vals := form["formats"]; len(vals) > 0 {
		formats = nil
		for _, v := range vals {
			for _, item := range strings.Split(v, ",") {
				item = strings.ToLower(strings.TrimSpace(item))
				if item != "" {
					formats = append(formats, item)
				}
			}
		}
		if len(formats) == 0 {
			formats = []string{"txt"}
		}
	}
	get := func(key string) string {
		if vs := form[key]; len(vs) > 0 {
			return strings.TrimSpace(vs[0])
		}
		return ""
	}
	return service.OcrOptions{
		Mode:      get("mode"),
		Quality:   get("quality"),
		Charset:   get("charset"),
		Lang:      get("lang"),
		Engine:    get("engine"),
		Redbox:    parseOcrBool(get("redbox")),
		Optimize:  parseOcrBool(get("optimize")),
		Translate: parseOcrBool(get("translate")),
		Exam:      parseOcrBool(get("exam")),
		Formats:   formats,
	}
}

func ocrURLMsg(err error) string {
	switch {
	case errors.Is(err, service.ErrBadURL):
		return "链接无效或不允许访问"
	case errors.Is(err, service.ErrTooLarge):
		return "链接文件超过大小限制"
	case errors.Is(err, service.ErrNotDoc), errors.Is(err, service.ErrNotImage):
		return "链接内容不是图片或 PDF"
	}
	return "文档链接读取失败"
}

// registerOcrOutputs turns script output files into stable download URLs.
func (h *OcrH) registerOcrOutputs(files map[string]string) []map[string]interface{} {
	out := []map[string]interface{}{}
	for _, key := range ocrExportOrder {
		path, ok := files[key]
		if !ok || path == "" {
			continue
		}
		url, err := h.Store.Register(path, ocrDownloadNames[key])
		if err != nil {
			fmt.Fprintf(os.Stderr, "[OCR] register %s failed: %v\n", key, err)
			continue
		}
		out = append(out, map[string]interface{}{
			"name": ocrDownloadNames[key],
			"type": key,
			"url":  url,
		})
	}
	return out
}

// HandleOcr: POST /api/ocr — accepts an uploaded image/PDF or document URL,
// starts recognition in the background, and returns {"success", "task_id"}.
// Progress is polled via GET /api/ocr/task/{id}.
func (h *OcrH) HandleOcr(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		h.writeErr(w, http.StatusMethodNotAllowed, "仅支持POST请求")
		return
	}

	tmp, err := h.newTmp()
	if err != nil {
		h.writeErr(w, http.StatusInternalServerError, "服务器错误")
		return
	}

	outdir := filepath.Join(tmp, "out")
	if err := os.MkdirAll(outdir, 0o755); err != nil {
		h.cleanupTmp(tmp)
		h.writeErr(w, http.StatusInternalServerError, "服务器错误")
		return
	}

	var srcPath, ext string
	source := "file"
	if isMultipartForm(r) {
		srcPath, ext, err = saveUpload(r, h.Cfg, "file", ocrExts, tmp)
		if err != nil {
			// Browsers and curl often send the document URL through a
			// multipart form too, so fall back to it when no file was posted.
			urlStr := strings.TrimSpace(r.PostFormValue("url"))
			if urlStr == "" {
				h.cleanupTmp(tmp)
				h.writeErr(w, http.StatusBadRequest, err.Error())
				return
			}
			srcPath, ext, err = service.FetchFile(tmp, urlStr, h.Cfg.MaxSize, ocrExts)
			if err != nil {
				h.cleanupTmp(tmp)
				h.writeErr(w, http.StatusBadRequest, ocrURLMsg(err))
				return
			}
			source = "url"
		}
	} else {
		if err := r.ParseForm(); err != nil {
			h.cleanupTmp(tmp)
			h.writeErr(w, http.StatusBadRequest, "参数格式错误")
			return
		}
		urlStr := strings.TrimSpace(r.FormValue("url"))
		if urlStr == "" {
			h.cleanupTmp(tmp)
			h.writeErr(w, http.StatusBadRequest, "请选择图片或 PDF 文件，或填写文档链接")
			return
		}
		srcPath, ext, err = service.FetchFile(tmp, urlStr, h.Cfg.MaxSize, ocrExts)
		if err != nil {
			h.cleanupTmp(tmp)
			h.writeErr(w, http.StatusBadRequest, ocrURLMsg(err))
			return
		}
		source = "url"
	}

	opts := parseOcrOptions(r)
	if err := opts.Normalize(); err != nil {
		h.cleanupTmp(tmp)
		h.writeErr(w, http.StatusBadRequest, err.Error())
		return
	}

	if h.Jobs == nil {
		h.cleanupTmp(tmp)
		h.writeErr(w, http.StatusInternalServerError, "服务器错误")
		return
	}
	job := h.Jobs.Create()
	if !h.Jobs.AcquireSlot() {
		h.Jobs.Delete(job.ID)
		h.cleanupTmp(tmp)
		h.writeErr(w, http.StatusServiceUnavailable, "服务器繁忙，请稍后重试")
		return
	}
	go func() {
		defer h.Jobs.ReleaseSlot()
		defer h.cleanupTmp(tmp)
		result, err := service.OcrFile(srcPath, opts, outdir)
		if err != nil {
			fmt.Fprintf(os.Stderr, "[OCR %s] error: %v\n", job.ID, err)
			h.Jobs.SetError(job.ID, "识别失败，请稍后重试")
			return
		}
		h.Jobs.SetComplete(job.ID, h.buildOcrPayload(result, opts, source, ext))
	}()

	h.writeJSON(w, http.StatusOK, map[string]interface{}{
		"success": true,
		"task_id": job.ID,
	})
}

func (h *OcrH) buildOcrPayload(result *service.OcrResult, opts service.OcrOptions, source, ext string) map[string]interface{} {
	text := result.Text
	return map[string]interface{}{
		"success":          true,
		"text":             text,
		"translated_text":  result.TranslatedText,
		"downloads":        h.registerOcrOutputs(result.Files),
		"engine":           result.Engine,
		"pages":            result.Pages,
		"chars":            result.Chars,
		"charset":          result.Charset,
		"from_text_layer":  result.FromTextLayer,
		"layout":           result.Layout,
		"blocks":           result.Blocks,
		"lines":            result.Lines,
		"filtered_headers": result.FilteredHeaders,
		"warnings":         result.Warnings,
		"elapsed_ms":       result.ElapsedMs,
		"options":          opts,
		"source":           source,
		"ext":              ext,
		"empty":            strings.TrimSpace(text) == "",
	}
}

File: src/handler/ocr.rs
use std::collections::HashMap;
use std::path::PathBuf;

use axum::extract::{Multipart, Path, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Json;
use serde_json::json;

use crate::service::ocr::{self, OcrOptions};
use crate::service::fetch_doc;
use crate::util::new_id;
use crate::AppState;

const OCR_EXTS: &[&str] = &["jpg", "jpeg", "png", "bmp", "tiff", "tif", "webp", "gif", "pdf"];
const DOWNLOAD_NAMES: [(&str, &str); 5] = [
    ("txt", "识别结果.txt"),
    ("docx", "识别结果.docx"),
    ("xlsx", "识别结果.xlsx"),
    ("json", "识别结果.json"),
    ("redbox", "红框标记图.png"),
];

fn is_true(v: &str) -> bool {
    matches!(
        v.trim().to_lowercase().as_str(),
        "1" | "true" | "yes" | "on"
    )
}

fn bad(msg: &str) -> Response {
    (
        StatusCode::BAD_REQUEST,
        Json(json!({ "success": false, "error": msg })),
    )
        .into_response()
}

fn server_error() -> Response {
    (
        StatusCode::INTERNAL_SERVER_ERROR,
        Json(json!({ "success": false, "error": "服务器错误" })),
    )
        .into_response()
}

/// Read multipart fields: at most one file plus repeatable text fields.
async fn read_multipart(
    multipart: &mut Multipart,
) -> Result<(Option<(Vec<u8>, String)>, HashMap<String, Vec<String>>), String> {
    let mut file: Option<(Vec<u8>, String)> = None;
    let mut fields: HashMap<String, Vec<String>> = HashMap::new();

    while let Some(mut field) = multipart
        .next_field()
        .await
        .map_err(|e| format!("请求格式错误: {e}"))?
    {
        let name = field.name().unwrap_or("").to_string();
        if let Some(file_name) = field.file_name() {
            if file.is_none() {
                let file_name = file_name.to_string();
                let mut buf = Vec::new();
                while let Some(chunk) = field
                    .chunk()
                    .await
                    .map_err(|e| format!("读取文件失败: {e}"))?
                {
                    buf.extend_from_slice(&chunk);
                }
                let ext = std::path::Path::new(&file_name)
                    .extension()
                    .map(|e| e.to_string_lossy().to_lowercase())
                    .unwrap_or_default();
                file = Some((buf, ext));
            }
            continue;
        }
        if let Ok(text) = field.text().await {
            fields.entry(name).or_default().push(text.trim().to_string());
        }
    }
    Ok((file, fields))
}

fn sanitize_ext(ext: &str) -> String {
    let lower = ext.to_lowercase();
    if OCR_EXTS.contains(&lower.as_str()) {
        lower
    } else {
        String::new()
    }
}

fn parse_options(fields: &HashMap<String, Vec<String>>) -> OcrOptions {
    let get = |key: &str| -> String {
        fields
            .get(key)
            .and_then(|vals| vals.first().cloned())
            .unwrap_or_default()
    };
    let mut formats: Vec<String> = Vec::new();
    if let Some(vals) = fields.get("formats") {
        for v in vals {
            for item in v.split(',') {
                let item = item.trim().to_lowercase();
                if !item.is_empty() {
                    formats.push(item);
                }
            }
        }
    }
    OcrOptions {
        mode: get("mode"),
        quality: get("quality"),
        charset: get("charset"),
        lang: get("lang"),
        engine: get("engine"),
        redbox: is_true(&get("redbox")),
        optimize: is_true(&get("optimize")),
        translate: is_true(&get("translate")),
        formats,
    }
}

/// POST /api/ocr — starts async recognition and returns {"success","task_id"}.
pub async fn handle_ocr(State(app): State<AppState>, mut multipart: Multipart) -> Response {
    let cfg = app.config.clone();
    let (file, fields) = match read_multipart(&mut multipart).await {
        Ok(v) => v,
        Err(e) => return bad(&e),
    };

    let tmp = cfg.tmp_dir.join(format!("ocr_{}", new_id(8)));
    let outdir = tmp.join("out");
    if std::fs::create_dir_all(&outdir).is_err() {
        return server_error();
    }

    let (src_path, ext, source) = if let Some((bytes, file_ext)) = file {
        let safe = sanitize_ext(&file_ext);
        if safe.is_empty() {
            return bad("不支持的文件类型，请上传图片或 PDF");
        }
        let path = tmp.join(format!("upload.{safe}"));
        if std::fs::write(&path, &bytes).is_err() {
            return server_error();
        }
        (path, safe, "file")
    } else if let Some(url) = fields.get("url").and_then(|v| v.first()) {
        if url.trim().is_empty() {
            return bad("请选择图片或 PDF 文件，或填写文档链接");
        }
        match fetch_doc(
            tmp.to_str().unwrap(),
            url.trim(),
            cfg.max_size,
            OCR_EXTS,
        )
        .await
        {
            Ok((path, file_ext)) => (PathBuf::from(path), file_ext, "url"),
            Err(e) => return bad(&e),
        }
    } else {
        return bad("请选择图片或 PDF 文件，或填写文档链接");
    };

    let mut opts = parse_options(&fields);
    if let Err(e) = ocr::normalize(&mut opts) {
        return bad(&e);
    }

    let job = app.ocr_jobs.create();
    let job_id = job.id.clone();
    if !app.ocr_jobs.acquire_one_slot() {
        app.ocr_jobs.delete(&job_id);
        return (
            StatusCode::SERVICE_UNAVAILABLE,
            Json(json!({ "success": false, "error": "服务器繁忙，请稍后重试" })),
        )
            .into_response();
    }

    let file_store = app.file_store.clone();
    let ocr_jobs = app.ocr_jobs.clone();
    let src_str = src_path.to_string_lossy().to_string();
    let out_str = outdir.to_string_lossy().to_string();
    let opts_json = serde_json::to_value(&opts).unwrap_or(serde_json::Value::Null);
    let opts_cloned = opts.clone();
    let source_cloned = source.to_string();
    let ext_cloned = ext.clone();

    tokio::spawn(async move {
        let (res, run_err) = match tokio::task::spawn_blocking(
            move || ocr::ocr_file(&src_str, &opts_cloned, &out_str),
        )
        .await
        {
            Ok(Ok(r)) => (Some(r), None),
            Ok(Err(e)) => (None, Some(e)),
            Err(e) => (None, Some(format!("OCR 任务异常: {e}"))),
        };

        let payload = match res {
            Some(r) => {
                let mut downloads = Vec::new();
                for (key, name) in DOWNLOAD_NAMES.iter() {
                    let Some(path) = r.files.get(*key) else { continue };
                    match file_store.register(path, name) {
                        Ok(url) => downloads.push(json!({
                            "name": name,
                            "type": key,
                            "url": url,
                        })),
                        Err(e) => tracing::warn!("OCR 下载注册失败 {key}: {e}"),
                    }
                }
                Some(json!({
                    "success": true,
                    "text": r.text,
                    "translated_text": r.translated_text,
                    "downloads": downloads,
                    "engine": r.engine,
                    "pages": r.pages,
                    "chars": r.chars,
                    "charset": r.charset,
                    "from_text_layer": r.from_text_layer,
                    "layout": r.layout,
                    "blocks": r.blocks,
                    "lines": r.lines,
                    "filtered_headers": r.filtered_headers,
                    "warnings": r.warnings,
                    "elapsed_ms": r.elapsed_ms,
                    "options": opts_json,
                    "source": source_cloned,
                    "ext": ext_cloned,
                    "empty": r.text.trim().is_empty(),
                }))
            }
            None => None,
        };

        let _ = std::fs::remove_dir_all(&tmp);
        ocr_jobs.release_one_slot();

        match payload {
            Some(p) => ocr_jobs.set_complete(&job_id, p),
            None => {
                tracing::error!("[OCR {job_id}] error: {}", run_err.unwrap_or_default());
                ocr_jobs.set_error(&job_id, "识别失败，请稍后重试");
            }
        }
    });

    Json(json!({ "success": true, "task_id": job.id }))
        .into_response()
}

/// GET /api/ocr/task/{id} — status and result of an async OCR job.
pub async fn handle_ocr_task(
    State(app): State<AppState>,
    Path(id): Path<String>,
) -> Response {
    let id = id.trim_matches('/');
    if id.is_empty() {
        return bad("缺少任务ID");
    }
    let Some(job) = app.ocr_jobs.get(id) else {
        return (
            StatusCode::NOT_FOUND,
            Json(json!({ "success": false, "error": "任务不存在或已过期" })),
        )
            .into_response();
    };
    Json(json!({
        "success": true,
        "status": job.status,
        "error": job.error,
        "result": job.result,
    }))
        .into_response()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_true_spellings() {
        for v in ["1", "true", "True", "YES", "on", " on "] {
            assert!(is_true(v));
        }
        for v in ["", "0", "false", "no", "off", "x"] {
            assert!(!is_true(v));
        }
    }

    #[test]
    fn test_sanitize_ext() {
        assert_eq!(sanitize_ext("PDF"), "pdf");
        assert_eq!(sanitize_ext("png"), "png");
        assert_eq!(sanitize_ext("exe"), "");
        assert_eq!(sanitize_ext("../etc"), "");
        assert_eq!(sanitize_ext(""), "");
    }

    #[test]
    fn test_parse_options_collects_formats() {
        let mut fields: HashMap<String, Vec<String>> = HashMap::new();
        fields.insert("mode".into(), vec!["advanced".into()]);
        fields.insert("formats".into(), vec!["docx,redbox".into(), "json".into()]);
        fields.insert("translate".into(), vec!["1".into()]);
        let o = parse_options(&fields);
        assert_eq!(o.mode, "advanced");
        assert_eq!(o.formats, vec!["docx", "redbox", "json"]);
        assert!(o.translate);
        assert!(!o.redbox);
    }

    #[test]
    fn test_parse_options_defaults_empty() {
        let fields: HashMap<String, Vec<String>> = HashMap::new();
        let o = parse_options(&fields);
        assert!(o.mode.is_empty());
        assert!(o.formats.is_empty());
    }
}
